### ISSUE TITLE
isolating Python in conda environments from site-packages

### ISSUE THREAD

**Issue by itcarroll** [conda/conda#13337]
### Checklist

- [X] I added a descriptive title
- [X] I searched open requests and couldn't find a duplicate

### What is the idea?

I am sorry to create a new issue for an old feature request, but it appears that previous issues requesting the same feature have been closed without resolution:

- #394
- #448
- #7707

The core problem is that the search path Python uses for packages is configurable, but conda does not leverage that to remove search paths that are outside the conda environment. Many users find the resulting behavior unintuitive, and it is different from the default behavior of Python's venv library for virtual environment creation. So this feature request has two parts

1. Environment creation needs an argument (default or not, I don't care) analagous to [`--system-site-packages`](https://peps.python.org/pep-0405/#isolation-from-system-site-packages).
2. Do it without using PYTHONNOUSERSITE environment variable.

The second part refers to current workarounds including the [conda-ecosystem-user-package-isolation](https://anaconda.org/conda-forge/conda-ecosystem-user-package-isolation) package or setting `PYTHONNOUSERSITE` manually. Here is a case where this workaround is not useful.

A command line tool installed outside a given conda environment using `pip install --user`, goes in `~/.local/bin/` and includes as its first line `#!/path/to/a/python`. *THAT* Python is outside the conda environment, but if `PYTHONNOUSERSITE` is active for the conda environment, site packages are disabled for *ALL* Python interpreters. This breaks the command line tool when invoked inside the conda environment.

Although conda does not create a *virtual* environment (as far as I understand ... and I may not), a `pyvenv.cfg` file is checked by the `site` module regardless. So, conda can do what venv does and create a `pyvenv.cfg` file with the line `include-system-site-packages = false`. I don't know what other repercussions that has involving `sys.prefix`, but so far so good when I do it manually. Needs expert input.

**Comment by travishathaway** [comment]
@itcarroll,

Please see my pull request:

- https://github.com/conda/conda/pull/13635

This integrates this feature a little more tightly into conda than what was done in your proposed plugin (https://github.com/conda-forge/staged-recipes/pull/24506)

**Comment by itcarroll** [comment]
I think this looks fantastic, thank you.

It looks like you get the "--no-python-user-site-packages" on `conda create`, `conda install`, and `conda update`. Is there any other way to set it as a global default (i.e. within .condarc) or as part of an environment definition file (i.e. environment.yml)?

I'm closing my PR on conda-forge in deference to this most excellent solution.

**Comment by travishathaway** [comment]
@itcarroll,

Yes, you can set it in the `.condarc` file:

```yaml
no_python_user_site_packages: true
```

or even as an environment variable:

```bash
export CONDA_NO_PYTHON_USER_SITE_PACKAGES=true
```

**Comment by mfansler** [comment]
@travishathaway looks excellent!

Just yesterday I was chatting with a senior Python user who was taken aback that Conda doesn't default to this like `venv` does. It seems like such a serious oversight given how much trouble and confusion user sites cause over the years. I'd strongly advocate that, if not this release, some future release make `True` the default setting.

**Comment by travishathaway** [comment]
@mfansler,

I agree with that, but we try to be conservative when introducing changes that could potentially disturb how environments function.

@jezdez and @kenodegard,

Could we use our deprecation policy to announce and plan for the default value of a setting? I'm not sure if we have used the deprecation mechanism in this way before, but it seems like it could be useful for giving our users advanced warning of a such a change.

**Comment by jezdez** [comment]
@travishathaway It's a good question, what you're describing is changing a core behavior, with a decade long history and strong opposition by the previous conda maintainer. As a former virtualenv author, I'm aware that isolation is the best course of action because it follows the principle of least surprise. I would strongly encourage you to read the tickets linked above carefully to understand the repurcusions. The core concern in the past was that conda is Python-agnostic, so it should not worry about a virtualenv related functionality. In the meantime the site module has been retrofitted in Python itself though, so I think it's good we review this past decission.

This isn't a deprecation per se (which implies eventual removal of an API in favor of another solution, to gracefully migrate existing downstream developers). The deprecation policy CEP isn't best suited for that, or at least I've had the experience that user don't react well when using something like that for such cases, since they don't realize what an alleged deprecation means for them, as deprecation policies are usually targetted at _developers_, not end users.

So in effect, for end users this is basically a _new_ feature, which must be rolled out without breaking existing environments.

I do think it's possible to eventually make this the default though, so here's my suggestion to roll it out in a staggered manner:

1. `24.3.x` We add an `include-system-site-packages` CLI/config option (note the inversion of the setting name) defaulting to `True`.
   - `conda info` prints out information if the currently active environment is including system site packages (including PEP 370 like `~/.local/lib/python3.11/site-packages`)
   - the release notes will announce this as a special announcement with the roll-out plan
3. `24.5.x` prints a warning when `include-system-site-packages` is set to `True`, preparing users that we'll switch to `include-system-site-packages = False` soon and provide help how to migrate their environments to have that pyenv.cfg file (`conda env isloate` as a helper?)
4. `24.7.x` prints a louder warning with the note of the imminent change of the default, with a reminder of all environments that haven't been migrated
5. `24.9.x` switches to `include-system-site-packages = False` by default, but keeps the warnings and provides help
6. `25.3.x` we remove the warnings and finalize the change of behavior

**Comment by Delengowski** [comment]
I found this issue when searching for a solution in the opposite direction.

I recognize that the way python looks for packages is roughly.

1. sys.modules
2. $PWD
3. ~/.local/lib/python3.X/site-packages
4. <prefix>/lib/python3.X/site-packages
(Not sure where $PYTHONPATH goes in here as I don't ever touch it.)

At my job we have a series of system level conda envs that users activate to fit their needs. If they require special versions of in house tools then the directive is to `pip install --user`, we refrain from making user level conda environments because the production environments are large and home area space is at a premium. Further this is an air gapped system and have to manage all our approved conda-forge packages on a local artifactory instance, so what environments could be made is limited to what's been approved.

The way that python looks for modules to import makes this work wonderfully. In fact it works great for runnable modules too. Individuals can perform `python -m MODULE` and if in the user site dir then it is ran. If there are conflicts we instruct to uninstall the user site dir package or simply do `python -I/-s -m MODULE` to isolate. Even further, we can have different user site dirs by manipulating `PYTHONUSERBASE` within a wrapper process, we've written a custom conda run plugin to support this.

I can appreciate where this is coming from, given the behavior of venv, which I didn't know about. However I am pretty against this becoming default behavior for conda. The precedence is set.

I found this issue when looking for a way to have conda prioritize .local/bin which are resulted from entry points when doing a `pip install --user`. Does anyone know of way to support that using conda (without direct manipulations of $PATH) or is it unavailable?

**Comment by itcarroll** [comment]
Not stale!

The conversation here and the linked PR seemed to get hung up on (1) whether isolation should (eventually) be the default and (2) whether a python specific option is allowed on `conda install`.

In the meantime, the conda plugin system seems to have come a long way. I am wondering whether a plugin is now the right way to provide this option for users?

**Comment by travishathaway** [comment]
Thanks pinging us on this @itcarroll.

I still think that the best way to do this is via a configuration variable. It might be best to move this issue over to the [conda-pypi ](https://github.com/conda/conda-pypi) project now. I'll ping the other maintainers about this.
